sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -72,7 +73,10 @@ class EquivalentExpressions {
     val skip = expr.isInstanceOf[LeafExpression] ||
       // `LambdaVariable` is usually used as a loop variable, which can't be evaluated ahead of the
       // loop. So we can't evaluate sub-expressions containing `LambdaVariable` at the beginning.
-      expr.find(_.isInstanceOf[LambdaVariable]).isDefined
+      expr.find(_.isInstanceOf[LambdaVariable]).isDefined ||
+      // `PlanExpression` wraps query plan. To compare query plans of `PlanExpression` on executor,
+      // can cause error like NPE.
+      (expr.isInstanceOf[PlanExpression[_]] && TaskContext.get != null)
```

Comment:
Just for curiosity, does this issue happen in interpreted code path as well? e.g. we send `PlanExpression` to executor side and eval it, and hit NPE.

Reply:
IIUC,`EquivalentExpressions` is only used in the codegen mode now, e.g., `GenerateUnsafeProjection` uses this class in common subexpr elimination, but `InterpretedUnsafeProject does not elimnate common subexprs.

Reply:
Not sure I understand your question correctly. But PlanExpressions of a SparkPlan are evaluated and updated (e.g., ExecSubqueryExpression.updateResult) with values before a query begins to run. The values are kept in PlanExpression, and on executor side when to call eval of PlanExpression, it simply returns the kept value. I think we do not really evaluate a PlanExpression at executor side.

Reply:
ah got it, so the kept value is serialized and sent to executor side in interpreted code path.

Reply:
This issue also reminds me that it's better to always do codegen at driver side, even if whole-stage-codegen is false. We can investigate it later.

Reply:
Ok. Please let me know if you have some ideas later.